harness/determined/_info.py

```diff
@@ -0,0 +1,318 @@
+        self.container_rank = container_rank
+
+    def _to_file(self, path: str = "/run/determined/info/rendezvous.json") -> None:
+        """
+        to_file writes the RendezvousInfo to a well-known location in a Determined container.  This
+        is called automatically early in the lifetime of a container, so user code can always expect
+        this file to be written before user code runs.
+        """
+        with open(path, "w") as f:
+            json.dump(vars(self), f)
+
+    @classmethod
+    def _from_file(
+        cls,
+        path: str = "/run/determined/info/rendezvous.json",
+    ) -> Optional["RendezvousInfo"]:
+        if not os.path.exists(path):
+            return None
+        with open(path) as f:
+            return cls(**json.load(f))
+
+
+class TrialInfo:
+    def __init__(
+        self,
+        trial_id: int,
+        experiment_id: int,
+        trial_seed: int,
+        hparams: Dict[str, Any],
+        config: Dict[str, Any],
+        latest_batch: int,
+        trial_run_id: int,
+        debug: bool,
+        unique_port_offset: int,
+        inter_node_network_interface: Optional[str],
+    ):
+        self.trial_id = trial_id
+        self.experiment_id = experiment_id
+        self.trial_seed = trial_seed
+        self.hparams = hparams
+
+        # _config is private because it's not a stable API; as the experiment config version
+        # increases, the shape of the experiment config dict might change as the master is upgraded.
+        # As a result, we should strongly discourage users from reading the experiment config dict
+        # in their training code.  They should focus on the data field and the hparams field, which
+        # have always been and will always be user-defined.
+        self._config = config
+
+        # rb: These fields are private because I am pretty confident that we need to find better
+        # ways to pass them around the system.  But for now, they're passed in as environment
+        # variables and for now we have to be able to handle that.
+        # TODO: decide if we really want to track latest_batch for users or not.
+        self._latest_batch = latest_batch
+        # TODO: get rid of trial_run_id in favor of allocation id.
+        self._trial_run_id = trial_run_id
+        # TODO: decide if the experiment config is the right place for users to set a debug flag.
+        self._debug = debug
+        # TODO: is this derivable from the slot ids?
+        self._unique_port_offset = unique_port_offset
+        # TODO: Get rid of this in favor of launch layer configs?
+        self._inter_node_network_interface = inter_node_network_interface
+
+    @classmethod
+    def _from_env(cls) -> "TrialInfo":
+        assert "DET_TRIAL_ID" in os.environ, "must be run inside a Trial container"
+        experiment_config = json.loads(os.environ["DET_EXPERIMENT_CONFIG"])
+        return cls(
+            trial_id=int(os.environ["DET_TRIAL_ID"]),
```

Comment:
what if we had a class for all these DET environment variables? could read by defined key consts instead of arbitrary strings and also enforce consistent read types like json.loads or ints

Reply:
I'm a little confused... I think `ClusterInfo` _is_ the class for all of these environment variables.  What did you have in mind?

Reply:
I was picturing some kind of master class that contained all the environment variables we used and some getters (maybe setters) that would give consistent return types. if the environment variables should only be used in context of ClusterInfo and RendezvousInfo and TrialInfo and not standalone, then it's pointless. but like if i want just DET_TRIAL_ID in some other part of the harness, maybe it would be nice if there was a single API for getting it? like env.trial_id or something. basically never having to pass in an arbitrary environment variable string and casting it to a type if used outside of the ...Info objects.

Reply:
> like env.trial_id or something

lol like `info.trial.trial_id`?  That's like the stated purpose of this class.

But in general, you shouldn't be reading these values at random points in the harness; that's how you build fragile code that only runs inside of our cluster.  Which is exactly what we did, which is exactly why we still haven't shipped local training mode.  _Everything_ assumes it can access values that are only set when on the cluster.  There's no clean interface anywhere.

Reply:
but the current structure means to read a singular value you'd have to read a batch of them related to trialInfo or clusterInfo. but your explanation of why they shouldn't be read this way makes sense.